enhancements/microshift/microshift-storage-version-migrator.md

```diff
@@ -0,0 +1,253 @@
+the API Server preferred API Resources, (this api request needs to be done in
+legacy mode, see Risks and Mitigations). Each APIResource will contain a
+`StorageVersionHash` that represents a sha256 sum of `group/version/kind`
+truncated to 8 characters and then base64 encoded. This value is stored in a
+`StorageState` CR during discovery. They then are compared, if there is a
+difference between the `StorageVersionHash` in the discovery information and the
+one stored in the `StorageState` for that specific resource, a
+`StorageVersionMigration` CR is created by the trigger.
+
+On start the Migrator will create a lister for `StorageVersionMigration`
+resources, and wait for a migration request for a resource. Once a migration
+request is received, it will query all of the resources for that version of that
+API. It will then sequentially migrate them one by one. Migrating a resource
+means querying the object, altering the version and performing a noop update.
+The result of this action is stored in the status for the
+`StorageVersionMigration`.
+
+### Risks and Mitigations
+
+We do have some risks with the feature as a whole but the mitigations are pretty
+straight forward.
+
+1. A risk that we have with this feature is that currently there is no real
+   upstream solution for storage version migrations. The current implementation
+   is in alpha and it does seem like it will be revisited in the future.
+   Luckily, until there is more consensus upstream we can continue to use the
+   approach outlined here but we will need to keep in mind that we might need to
+   re-work this in the future.
+
+2. In that same vein, we do run the risk that the code doesn't get touched much
+   due to the nature and life of the current Storage Version Migrator solution.
+   So we might run into dependency problems in the future. The good news there
+   is that we have good ties to the upstream developers so the procedure for
+   resolving any future dependency problems would be the same as the other
+   services we run.
+
+3. Another risk is that the new [Aggregated
+   Discovery](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery)
+   removes the `StorageVersionHash`, luckily the change does take into account
+   that the Migrator uses it and we just need to make sure our [discovery client
+   is in legacy
+   mode](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3352-aggregated-discovery#is-the-rollout-accompanied-by-any-deprecations-andor-removals-of-features-apis-fields-of-api-types-flags-etc)
+
+4. Lastly, migrations can fail, so there can be situations when a migration
+   might fail on a specific resource. We will be able to catch these failures in
+   our own e2e tests for our resources. However, when it comes to user
+   resources, it is up to the users to update those resources to resolve the
+   failure.
```

Comment:
Do migration failures trigger a failed health check? Should they?

Reply:
They currently don't fail health checks, but I think they shouldn't. 

The Migrator will try it's best to update but if it can't get it, then human intervention is needed for that resource. It should be part of our guide that if you're running an upgrade, check to make sure you don't have failed migrations.

Plus, unless the user deletes all reference to the previous version and somehow doesn't blow out the resource that reference it in the process, a migration failure shouldn't put the cluster in an healthy state. Given what's stated about no upstream solution there is a lot of care given to removal of APIs in the core resources as well.

Reply:
OK, I'm thinking of behavior for layered products.

@ncdc @joelanford @stevekuznetsov we may want to think about how an OLM operator adds a health-check step, if it wants to. The original work on that for MicroShift itself is described in https://github.com/openshift/enhancements/blob/master/enhancements/microshift/microshift-greenboot.md

Reply:
We have a proposal for adding a health check to part of OLM v1. See https://docs.google.com/document/d/1MwCD4RINQYPwDENcTL-MlJQAmsrbAkE2togpaGGRLAM/edit for more details.